repo/repo.go

```diff
@@ -98,33 +91,94 @@ func ListRepositories(workspace *rule.File) (repos []*rule.Rule, repoFileMap map
 			if err != nil {
 				return nil, nil, err
 			}
+			leveled := strings.HasPrefix(f, "+")
+			f = strings.TrimPrefix(f, "+")
 			f = filepath.Join(filepath.Dir(workspace.Path), filepath.Clean(f))
-			macroFile, err := rule.LoadMacroFile(f, "", defName)
-			if err != nil {
+			checked[f+"%"+defName] = true
+
+			la := &loadArgs{
+				leveled: leveled,
+				workspace: workspace.Path,
+				f: f,
+				defName: defName,
+				repos: repos,
+				repoFileMap: repoFileMap,
+				repoIndexMap: repoIndexMap,
+				checked: checked,
+			}
+
+			if err := loadRepositoriesFromMacro(la); err != nil {
 				return nil, nil, err
 			}
-			for _, repo := range macroFile.Rules {
-				if name := repo.Name(); name != "" {
-					repos = append(repos, repo)
-					repoFileMap[name] = macroFile
-					repoIndexMap[name] = len(repos) - 1
+			repos = la.repos
+		}
+	}
+	return repos, repoFileMap, nil
+}
+
+type loadArgs struct {
+	leveled bool
+	workspace string
+	f string
+	defName string
+	repos []*rule.Rule
+	repoFileMap map[string]*rule.File
+	repoIndexMap map[string]int
+	checked map[string]bool
+}
+
+func loadRepositoriesFromMacro(la *loadArgs) error {
```

Comment:
I think the following function signature would be a little less error-prone:
```
func loadRepositoriesFromMacro(la *loadArgs, leveled bool, f, defName string) error
```
`loadArgs` are things that are shared between the recursive calls, whereas the explicitly passed in parameters change between every recursive call